Fix Giant's Knife unexpected behaviours

Sarge-117 opened the pull request:
These changes fix two unintended behaviours with the Giant's Knife.

1) (z_parameter.c changes) - in rando, if we break and re-buy the Giant's Knife without ever having picked up the Kokiri Sword, the Giant's Knife "broken" flag can end up out of sync with the Giant's Knife health.  I.E. you can end up with a sword that is functionally broken, but without the broken flag bring "true".  This will prevent Medigoron from selling you a new one, even though your sword is useless.

Footage: https://cdn.discordapp.com/attachments/988962246585634856/998697205898694707/giantsknife.mp4

The above scenario doesn't normally happen outside of rando, since the player should always have the Kokiri Sword when buying the Giant's Knife from Medigoron.

2) (z_kaleido_equipment.c changes) - fixes #812, where breaking the giant's knife would correctly update the flag and your B button, but was not drawing the "broken sword" icon on the inventory subscreen.

Sarge-117 committed:
Giant's Knife Behaviour Fixes

Fixes a case where Giant's Knife (specifically, breaking and re-buying it) can behave unexpectedly if you don't have a Kokiri Sword in your inventory.  Also fixes the broken icon not showing up in inventory after you break it.

Sarge-117 committed:
less whitespace

briaguya0 commented:
the video attached to #812 shows the kokiri sword in inventory, and the issue mentions build 192 (rando was merged in for build 197)

could you clarify what's causing this outside of rando?

Sarge-117 commented:
To be clear, the unintended behaviour can still occur if you are messing with the save editor or debug menu (for example, if you manually set your sword health >0, but leave the "broken" flag turned on).  You could also end up at Medigoron without the Kokiri Sword through glitches.  The fixes above should work fine for normal gameplay.

Sarge-117 commented:
> the video attached to #812 shows the kokiri sword in inventory, and the issue mentions build 192 (rando was merged in for build 197)
> 
> could you clarify what's causing this outside of rando?

The video in #812 is only related to the icon that is displayed on the inventory screen.  That issue was not dependent on the Kokiri Sword.  Having the Kokiri Sword is necessary is you break your Giant's Knife and then re-buy it from Medigoron.

briaguya0 commented:
so with this PR
* icons are working for both b-button and kaleidoscope (rando and vanilla)
* it's still possible to glitch your way into an out of sync broken giant's knife state in vanilla
* it's no longer possible to get into an out of sync broken giant's knife state in rando

correct?

Sarge-117 commented:
> so with this PR
> 
>     * icons are working for both b-button and kaleidoscope (rando and vanilla)
> 
>     * it's still possible to glitch your way into an out of sync broken giant's knife state in vanilla
> 
>     * it's no longer possible to get into an out of sync broken giant's knife state in rando
> 
> 
> correct?

Correct!  You could also still also get into the out of sync state through debug warping or whatever.

Sarge-117 committed:
Merge branch 'HarbourMasters:develop' into fix-broken-knife-inventory-icon

briaguya0 commented:
so it looks like the icon issue was probably introduced here https://github.com/HarbourMasters/Shipwright/commit/7d0af303ec8929c8d30bba8b41f90479eac152ac

i'm not sure what would break if we reverted this, but it seems we shouldn't need to do the complex check that this PR has to restore the broken sword icon

Sarge-117 commented:
> so it looks like the icon issue was probably introduced here [7d0af30](https://github.com/HarbourMasters/Shipwright/commit/7d0af303ec8929c8d30bba8b41f90479eac152ac)
> 
> i'm not sure what would break if we reverted this, but it seems we shouldn't need to do the complex check that this PR has to restore the broken sword icon

Thank you for mentioning that @briaguya-ai 

I decided to test going back & forth through time and discovered that the broken icon doesn't go grayscale as child, so it appears that this isn't quite ready yet.  I will mark as draft and do some more testing!

Sarge-117 committed:
fix 0xE check

Accidentally committed with the wrong check

vaguerant commented:
I think the problem here is that we're tying together "**grayscale and** draw textures for non-BGS/GK equipment" which isn't necessary. Checking the upstream decomp, this isn't even where grayscaling occurs normally, so it should be fine to just move it up a bit.

```c
        for (k = 0, bit = rowStart, point = 4; k < 3; k++, point += 4, temp++, bit++) {

            int itemId = ITEM_SWORD_KOKIRI + temp;
            bool not_acquired = (gItemAgeReqs[itemId] != 9) && (gItemAgeReqs[itemId] != gSaveContext.linkAge);
            if (not_acquired) {
                gsDPSetGrayscaleColor(POLY_KAL_DISP++, 109, 109, 109, 255);
                gsSPGrayscale(POLY_KAL_DISP++, true);
            }
            if (((u32)i == 0) && (k == 2) && (gSaveContext.bgsFlag != 0)) {
                KaleidoScope_DrawQuadTextureRGBA32(globalCtx->state.gfxCtx, gBiggoronSwordIconTex, 32, 32, point);
            } else if ((i == 0) && (k == 2) && (gBitFlags[bit + 1] & gSaveContext.inventory.equipment)) {
                KaleidoScope_DrawQuadTextureRGBA32(globalCtx->state.gfxCtx, gBrokenGiantsKnifeIconTex, 32, 32, point);
            } else if (gBitFlags[bit] & gSaveContext.inventory.equipment) {
                KaleidoScope_DrawQuadTextureRGBA32(globalCtx->state.gfxCtx, gItemIcons[itemId], 32, 32, point);
            }
            gsSPGrayscale(POLY_KAL_DISP++, false);
        }
```
This seems to solve everything on my end, with four possible outcomes: unbroken sword in color, unbroken sword in grayscale, broken sword in color, broken sword in grayscale. Is there anything I'm missing?

**EDIT:** While we're at it, maybe it's worth renaming `bool not_acquired`, since it actually tracks whether the item is restricted from being used at your current age.

Sarge-117 committed:
Move the grayscale code up

Moved the grayscaling portion up higher in the function and renamed "bool not_acquired" to "bool age_restricted".  Thanks to @vaguerant for the suggestion!

Sarge-117 commented:
Thank you @vaguerant!  I tested your suggestion as child and adult and in rando and non-rando save files.  It seems to work perfectly!  I also changed `bool not_acquired` to `bool age_restricted`, which seems more appropriate.

Sarge-117 commented:
Made this ready for review again - sorry for the confusion @Kenix3 @briaguya-ai

Sarge-117 committed:
Merge branch 'HarbourMasters:develop-zhora' into fix-broken-knife-inventory-icon

briaguya0 reviewed (approved):
:shipit:
